File: android-app/src/main/kotlin/com/jworks/kanjiquest/android/ui/subscription/SubscriptionViewModel.kt
package com.jworks.kanjiquest.android.ui.subscription

import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import com.jworks.kanjiquest.core.domain.UserSessionProvider
import com.jworks.kanjiquest.core.domain.model.Subscription
import com.jworks.kanjiquest.core.domain.model.SubscriptionPlan
import com.jworks.kanjiquest.core.domain.repository.AuthRepository
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.launch
import javax.inject.Inject

data class SubscriptionUiState(
    val isLoggedIn: Boolean = false,
    val isPremium: Boolean = false,
    val subscription: Subscription? = null,
    val isLoading: Boolean = true
)

@HiltViewModel
class SubscriptionViewModel @Inject constructor(
    private val authRepository: AuthRepository,
    private val userSessionProvider: UserSessionProvider
) : ViewModel() {

    private val _uiState = MutableStateFlow(SubscriptionUiState())
    val uiState: StateFlow<SubscriptionUiState> = _uiState.asStateFlow()

    init {
        loadSubscription()
    }

    private fun loadSubscription() {
        viewModelScope.launch {
            val userId = authRepository.getCurrentUserId()
            val subscription = authRepository.getSubscription()
            _uiState.value = SubscriptionUiState(
                isLoggedIn = userId != null,
                isPremium = subscription?.plan == SubscriptionPlan.PREMIUM,
                subscription = subscription,
                isLoading = false
            )
        }
    }

    fun refresh() {
        _uiState.value = _uiState.value.copy(isLoading = true)
        loadSubscription()
    }
}

File: ios-app/KanjiQuest/Screens/Subscription/SubscriptionViewModel.swift
import Foundation
import SharedCore

/// Subscription view model. Mirrors Android's SubscriptionViewModel.kt.
@MainActor
class SubscriptionViewModel: ObservableObject {
    @Published var isLoggedIn = false
    @Published var isPremium = false
    @Published var isLoading = true

    func load(container: AppContainer) {
        Task {
            let userId = try? await container.authRepository.getCurrentUserId()
            let subscription = try? await container.authRepository.getSubscription()
            isLoggedIn = userId != nil
            isPremium = subscription?.plan == .premium
            isLoading = false
        }
    }

    func refresh(container: AppContainer) {
        isLoading = true
        load(container: container)
    }
}
